## Импортируем и читаем какие-то непонятные буковки и циферки
Первым делом нужно понять, с чем мы имеем дело. Открываем наш датасет и читаем его. 

In [47]:
import csv

with open('disney_plus_shows.csv', 'r', encoding='utf-8') as f:
    csv_reader = csv.DictReader(f)
    
    data = list(csv_reader)
#print(data)

## Сколько же всего фильмов? Хватит, чтобы съесть весь попкорн?

Начнем с базы. Мы  узнаем сколько всего контента: Чего больше? Фильмов или же сериалов?)


In [17]:
# 1. Подсчет фильмов и сериалов
films = 0
serials = 0
no_data = 0
# Проходим по всем элементам
for kontent in data:
    if kontent['type'] == 'movie':
        films += 1
    elif kontent['type'] == 'series':
        serials += 1
        
# Выводим результаты
print("Всего фильмов:", films)
print("Всего сериалов:", serials)
print("Всего контента:", films + serials)

Всего фильмов: 680
Всего сериалов: 191
Всего контента: 871


 ## ТОП-10 фильмов с самыми высокими рейтингами IMDb.
Как истиный фанат франшизы Disney, вы встаете перед выбором, что посмотреть вечером выходного и решаете выбрать самый лучший фильм по  рейтингу IMDb, поэтому вот вам топ фильмов на вечер)

In [16]:
movie_ratings = {} #создаем новый словарь

for item in data: # проходим по всем элементам в data
    if item['type'] == 'movie' and item['imdb_rating'] and item['imdb_rating'] != 'N/A': # проверяем, что нет пустых значений, они на не нужныыы
        movie_ratings[item['title']] = item['imdb_rating'] # добавляем в новый словарь ключи и значения.

movies_for_sorting = [] #создаем пустой список
for movie, rating in movie_ratings.items(): # для названия и рейтинга из словаря
    movies_for_sorting.append((rating, movie)) #создаем список с рейтингом и названием

movies_for_sorting.sort(reverse=True) #сортируем список и меняем его от большего к меньшему

top_10_movies = movies_for_sorting[:10] # берем тольк опервые 10 из отсортированного списка

print("ТОП-10 фильмов по рейтингу IMDb:") # выводим результат
for i in range(10):
    movie_info = top_10_movies[i]
    print('№',i+1, movie_info[1],"-", movie_info[0])

ТОП-10 фильмов по рейтингу IMDb:
№ 1 Star Wars: Episode V - The Empire Strikes Back - 8.7
№ 2 Star Wars: Episode IV - A New Hope - 8.6
№ 3 The Lion King - 8.5
№ 4 Disney's Newsies: The Broadway Musical! - 8.5
№ 5 WALL·E - 8.4
№ 6 Presto - 8.4
№ 7 Piper - 8.4
№ 8 Coco - 8.4
№ 9 Avengers: Endgame - 8.4
№ 10 Toy Story 3 - 8.3


## Какой самый продуктивный год?
Были ли у Disney годы, когда они не спали ночами, чтобы радовать нас новинками? Сейчас узнаем, какой год был самым активным.

In [22]:
# Собираем статистику по годам
# задача: составить словарь ,где ключем будет год, а значением - кол-во фильмов ,выпущенных в этот год.
year_status = {}
no_data = 0

for item in data:
    year = item['year'] # присваеваем году значение года из словаря. 
    
    if not year or year == 'N/A': # отбрасываем некоректные данные или же пустые 
        no_data += 1 # добавляем их к счетсику
    else:
        if year not in year_status: #проверяем. что его еще нет в словаре
            year_status[year] = 0 # если год встречается впервые, то создаем новый
        year_status[year] += 1 #встречается повторно, поэтому добавляем его к счетчику
sorted_year_status = sorted(year_status.items()) #сортируем список

max_count2 = 0

for year, count in year_status.items():
    if count > max_count2:
        max_count2 = count
        max_year2 = year

print('Самый продуктивный год:', max_year2, '-', max_count2, 'записей')

Самый продуктивный год: 2019 - 31 записей


## Какой самый длинный фильм? Какая средняя продолжительность фильмов/сериалов?
Хватит ли время фильма пока вы в дороге? Или же наоборот хонометраж окажется слишком маленьким(

In [46]:
# Создаем списки для хранения данных
movies_runtime = []
series_runtime = []

movies_titles = []
series_titles = []

# Обрабатываем каждый элемент данных
for item in data:
    # Проверяем тип и продолжительность (исключаем 'N/A' и пустые строки)
    if item['runtime'] != 'N/A' and item['runtime'] != '':
        # Извлекаем число из строки продолжительности
        runtime_parts = item['runtime'].split()
        if len(runtime_parts) > 0:  # Проверяем, что есть хотя бы одна часть
            runtime_number = int(runtime_parts[0])
            
            # Разделяем по типу
            if item['type'] == 'movie':
                movies_runtime.append(runtime_number)
                movies_titles.append(item['title'])
            elif item['type'] == 'series':
                series_runtime.append(runtime_number)
                series_titles.append(item['title'])

# Вычисляем среднюю продолжительность для фильмов
count_movies = len(movies_runtime)
total_movies = sum(movies_runtime)
avg_movie = total_movies / count_movies

# Вычисляем среднюю продолжительность для сериалов
count_series = len(series_runtime)
total_series = sum(series_runtime)
avg_series = total_series / count_series

# Для фильмов
longest_movie_time = max(movies_runtime)
shortest_movie_time = min(movies_runtime)
longest_movie_title = movies_titles[movies_runtime.index(longest_movie_time)]
shortest_movie_title = movies_titles[movies_runtime.index(shortest_movie_time)]

# Для сериалов
longest_series_time = max(series_runtime)
shortest_series_time = min(series_runtime)
longest_series_title = series_titles[series_runtime.index(longest_series_time)]
shortest_series_title = series_titles[series_runtime.index(shortest_series_time)]

# Вывод результатов
print('Средняя продолжительность фильмов:', avg_movie, 'мин')
print('Средняя продолжительность сериалов:', avg_series, 'мин')
print()
print('Самый длинный фильм:', longest_movie_title, '-', longest_movie_time, 'мин')
print('Самый короткий фильм:', shortest_movie_title, '-', shortest_movie_time, 'мин')
print()
print('Самый длинный сериал(одна серия):', longest_series_title, '-', longest_series_time, 'мин')
print('Самый короткий сериал(одна серия):', shortest_series_title, '-', shortest_series_time, 'мин')

Средняя продолжительность фильмов: 78.38670694864048 мин
Средняя продолжительность сериалов: 28.16883116883117 мин

Самый длинный фильм: Avengers: Endgame - 181 мин
Самый короткий фильм: For the Birds - 3 мин

Самый длинный сериал(одна серия): Dog Whisperer with Cesar Millan - 110 мин
Самый короткий сериал(одна серия): Star Wars Blips - 1 мин


## Кто же больше всех потрудился?
Какой же режиссер стал самым продуктивным?

In [12]:
# Собираем режиссеров
directors = {} #создаем пустой словарь
for item in data: #перебираем каждый элемент в данных
    if item['director']: 
        director_list = item['director'].split(',') #разбиваем по запятым, режиссеров много мб
        for director in director_list: # перебираем каждого режиссера 
            if director not in directors: #проверяем есть ли этот режиссер в нашем списке
                directors[director] = 0 # если режиссера нет, добавляем с начальным значением 0
            directors[director] += 1 # если уже есть, то добавляем его к счетчику

# Создаем список для сортировки
directors_list = []
for director_name in directors:
    directors_list.append([directors[director_name], director_name])

# Сортируем и переворачиваем
directors_list.sort()
directors_list.reverse()

print("Топ режиссеров по количеству работ:") 
for director_info in directors_list[:10]: # выводим только первые топ 10
    print(f"{director_info[1]}: {director_info[0]} работ") # выводим данные

Топ режиссеров по количеству работ:
N/A: 205 работ
Jack Hannah: 16 работ
Paul Hoen: 15 работ
Wilfred Jackson: 12 работ
John Lasseter: 12 работ
James Algar: 12 работ
Clyde Geronimi: 12 работ
 Hamilton Luske: 12 работ
Robert Stevenson: 11 работ
Vincent McEveety: 10 работ


### Мдээээ, увы ,видимно не хватает данных по режисерам. Более 20% фильмов без данных, но ладно. Самый большой трудяга по этим данным: Джек Ханна, режиссер достаточно старых фильмов Disney. В основном его работы 40-х и 50-х годов.

## Какой жанр стал главным в киноиндустрии?
Пишем дисс на Дисней. Оказывается, у них есть явные жанры - фавориты.

In [11]:
# Собираем режиссеров
genres = {} #создаем пустой словарь
for item in data: #перебираем каждый элемент в данных
    if item['genre']: 
        genre_list = item['genre'].split(',') #разбиваем по запятым, вдруг несколько жанров
        for genre in genre_list: # еребираем каждого режисера 
            if genre not in genres: #проверяем естть ли этот режиссер в нашем списке
                genres[genre] = 0 # если режисера нет, добавляем с начальным значением 0
            genres[genre] += 1 # если уже есть, то добавлем его к счетчику

genres_list = []
for genre_name in genres:
    genres_list.append([genres[genre_name], genre_name])

# Сортируем и переворачиваем
genres_list.sort()
genres_list.reverse()

print("Топ жанров по количеству работ:") 
for genre_info in genres_list[:10]: # выводим только первые топ 10
    print(f"{genre_info[1]}: {genre_info[0]} работ") # выводим данные

Топ жанров по количеству работ:
 Family: 619 работ
Animation: 356 работ
 Comedy: 333 работ
 Adventure: 265 работ
 Fantasy: 241 работ
Comedy: 163 работ
 Sci-Fi: 153 работ
 Drama: 146 работ
 Short: 130 работ
 Musical: 120 работ


## Барабанная дробь... ИТОГИ ПОГРУЖЕНИЯ ВО ВСЕЛЕННУЮ DISNEY

Итого, используя открытые данные с сайта Kaggle мы смогли понять, что:
1. Фильмов больше, чем сериалов
2.  Самым высокорейтинговый фильм - **Звёздные войны: Эпизод 5 – Империя наносит ответный удар** (Фанаты Звездных войн, вы рады?).
3.  Больше всего контента Disney выпускал в **2019 году - целых 31!!**
4.  Самым активным режисером стал **Джек Ханна**.
5.  Самым численным жанром франшизы стал **семейный фильм**, что неудивительно для вселенной Disney, так как в основном он ориентирован на аудиторию разных возрастов и полов, с акцентом на семейную группу с детьми.
6.  Средняя длина фильмов составляет **78 минут**, в то время как сериалы длятся в среднем **28 минут** за эпизод
7.  Самый длинный фильм в библиотеке - **«Avengers: Endgame» (181 минута)**
8.  Самый короткий фильм - **«For the Birds» (окак, всего 3 минуты)**
9.  Среди сериалов рекордсменом по длине эпизода стал **«Dog Whisperer with Cesar Millan» (110 минут)**
10.  Самый короткий эпизод сериала - **«Star Wars Blips» (1 минута)**

In [3]:
with open('castle.txt', encoding='utf-8') as f:
    castle = f.read()
print(castle)

                                           %                                                                            
                                           @                                                                            
                                  =                                                                                     
                                  @      -    %                                                                         
                                # #     @   #=                                                                          
                            #:   =: +  @   %                                                                            
                              @     % -  ==   =%                                                                        
                                 %  @ @ #            =                  .@#                                             
                              . 